### DISCUSSION TITLE
How to configure apache-airflow-providers-apache-kafka with Kerberos?

### DISCUSSION THREAD

**Discussion by makrushin-evgenii** [apache/airflow#35148] [Q&A] [answered]
First I installed `apache-airflow-providers-apache-kafka==1.2.0`, add connection:
```
{
  "bootstrap.servers": "srs-d-kfk2-01.dev.kontur.ru:9093,srs-d-kfk2-02.dev.kontur.ru:9093,srs-d-kfk2-03.dev.kontur.ru:9093",
  "security.protocol": "SASL_PLAINTEXT",
  "sasl.kerberos.service.name": "kafka",
  "sasl.kerberos.principal": "svc_srs_khajiit@KONTUR",
  "sasl.kerberos.kinit.cmd": "kinit -R || {PASSWORD} | kinit svc_srs_khajiit@KONTUR"
}
```
and add some packages to Airflow image:
```
RUN apt-get update && \
    export DEBIAN_FRONTEND=noninteractive && \
    apt-get install -y \
        libsasl2-modules-gssapi-mit \
        libsasl2-dev \
        krb5-user && \
    rm -rf /var/lib/apt/lists/*
```

And get `Failed to create producer: No provider for SASL mechanism GSSAPI: recompile librdkafka with libsasl2 or openssl support. Current build options: PLAIN SASL_SCRAM OAUTHBEARER"` error. So i decided to rebuild `librdkafka` with libsasl2 or openssl support:
```
RUN apt-get update && \
    apt-get install -y \
        gcc \
        g++ \
        make \
        git-core  \
        patch \
        zlib1g-dev \
        libzstd-dev \
        libssl-dev \
        libsasl2-dev \
        libsasl2-modules-gssapi-mit \
        krb5-user
RUN git clone --depth 1 --branch v1.8.2 https://github.com/edenhill/librdkafka.git .
RUN ./configure --install-deps --disable-lz4-ext --enable-gssapi --enable-zstd --enable-zlib
RUN make -j
RUN make install
```

After this i still got same error. Probably `apache-airflow-providers-apache-kafka` bundled with its own `librdkafka` and ignore my. How to handle it?

**Comment by Taragolis** [comment]
> After this i still got same error. Probably apache-airflow-providers-apache-kafka bundled with its own librdkafka and ignore my.

You need to go deeper, Airflow doesn't bundle `librdkafka`, it is only use python dependencies listed in provider: https://airflow.apache.org/docs/apache-airflow-providers-apache-kafka/stable/index.html

Pre-build binaries comes through wheel distribution of [`confluent-kafka`](https://pypi.org/project/confluent-kafka/#files), so my assumption that you need to start from this point and build your own [wheel from source distribution](https://pip.pypa.io/en/stable/cli/pip_wheel/), by something like: 
```shell
python -m pip wheel confluent-kafka --use-pep517 --no-binary "confluent-kafka"
```

And after that install the resulting wheel package

**Reply by makrushin-evgenii** [reply]
Thank for your answer. It brought me very close to the solution

**Comment by makrushin-evgenii** [comment, marked as answer]
[Found this instuction](https://github.com/confluentinc/confluent-kafka-python/blob/master/INSTALL.md). It's not worked as is - specified repository not contains latest librdkafka versions. Solution is to build librdkafka from sources, then install confluent-kafka with `-no-binary` flag:
```
RUN apt update && \
    apt install -y gcc g++ make git-core patch zlib1g-dev libzstd-dev libssl-dev libsasl2-dev wget  \
    software-properties-common lsb-release python3 python3-pip python3-dev libsasl2-modules-gssapi-mit krb5-user
RUN git clone --depth 1 --branch v2.2.0 https://github.com/edenhill/librdkafka.git .
RUN ./configure --install-deps --disable-lz4-ext --enable-gssapi --enable-zstd --enable-zlib
RUN make -j && make install
RUN python3 -m pip install --no-binary confluent-kafka confluent-kafka
```
